test/adapters/apply-edit-adapter.test.ts

```diff
@@ -186,5 +189,235 @@ describe("ApplyEditAdapter", () => {
+      expect(fs.readFileSync(oldUri).toString()).to.equal('abcd')
+      expect(fs.existsSync(newUri)).to.equal(true)
+      expect(fs.readFileSync(newUri).toString()).to.equal('efgh')
+
+      expect(
+        (atom as any).notifications.addError.calledWith("workspace/applyEdits failed", {
+          description: "Failed to apply edits.",
+          detail: "Error during rename resource operation: Target exists.",
+        })
+      ).to.equal(true)
+    })
+
+    it("handles delete resource operations on files", async () => {
+      const directory = fs.mkdtempSync(tempDir)
+      const uri = path.join(directory, "test.txt")
+      fs.writeFileSync(uri, 'abcd')
+
+      const result = await ApplyEditAdapter.onApplyEdit({
+        edit: {
+          documentChanges: [
+            {
+              kind: "delete",
+              uri: uri
+            }
+          ]
+        },
+      })
+
+      expect(result.applied).to.equal(true)
+      expect(fs.existsSync(uri)).to.equal(false)
+    })
+
+    it("handles delete resource operations on directories", async () => {
+      const directory = fs.mkdtempSync(tempDir)
+      const file1 = path.join(directory, '1.txt')
+      const file2 = path.join(directory, '2.txt')
+      fs.writeFileSync(file1, '1')
+      fs.writeFileSync(file2, '2')
+
+      const result = await ApplyEditAdapter.onApplyEdit({
+        edit: {
+          documentChanges: [
+            {
+              kind: "delete",
+              uri: directory,
+              options: {
+                recursive: true
+              }
+            }
+          ]
+        },
+      })
+
+      expect(result.applied).to.equal(true)
+      expect(fs.existsSync(directory)).to.equal(false)
+      expect(fs.existsSync(file1)).to.equal(false)
+      expect(fs.existsSync(file2)).to.equal(false)
+    })
+
+    it("throws an error when deleting a non-empty directory without recursive option", async () => {
+      const directory = fs.mkdtempSync(tempDir)
+      const file1 = path.join(directory, '1.txt')
+      const file2 = path.join(directory, '2.txt')
+      fs.writeFileSync(file1, '1')
+      fs.writeFileSync(file2, '2')
+
+      const result = await ApplyEditAdapter.onApplyEdit({
+        edit: {
+          documentChanges: [
+            {
+              kind: "delete",
+              uri: directory,
+              options: {
+                recursive: false
+              }
+            }
+          ]
+        },
+      })
+
+      expect(result.applied).to.equal(false)
+      expect(fs.existsSync(directory)).to.equal(true)
+      expect(fs.existsSync(file1)).to.equal(true)
+      expect(fs.existsSync(file2)).to.equal(true)
+      const errorCalls = (atom as any).notifications.addError.getCalls()
+      expect(errorCalls.length).to.equal(1)
+      expect(errorCalls[0].args[1].detail).to.match(/Error during delete resource operation: (.*)/)
+    })
+
+
+    it("throws an error on delete operation if target doesnt exist", async () => {
```

Comment:
Should the delete operation be idempotent and fail silently if the file has already been deleted?

Reply:
I'm not sure.
LSP's `CreateFile` actually has an `ignoreIfNotExists` param.
What I implemented is:
- If `ignoreIfNotExists` is `undefined`, assume `false` and throw error
- If `ignoreIfNotExists` is `true`, return in case the file (or directory) doesn't exist

All ResourceOperations of such kind of params, notably `overwrite`, `ignoreIfExists` and `ignoreIfNotExists`.
I don't think that LSP gives any guidance on what should be considered the default value for such params, so I just assumed `undefined === false`, but I agree it might make sense to adapt with more sensible defaults...
WDYT?

Reply:
@aminya what do you think? I am on the fence on this one. On the one hand it makes sense to have `undefined` == `false`. On the other hand I don't see any situation where someone would want an error when deleting a file if the file already doesn't exist.

Reply:
I actually agree that a delete operation should not throw an error if the file doesn't exist.
For all other params/operation the sensible default would be `false` like : 
- A `rename` operation should not overwrite another file and fail by default if the target exists (defaults: `overwrite = false` , `ignoreIfExists = false`
- A `create` operation should not overwrite another file and fail by default if the target exists (defaults: `overwrite = false` , `ignoreIfExists = false`
- A `delete` operation should not fail by default if the target doesn't exist (defaults: `ignoreIfNotExists = true` **<== our case here**

So in case of a delete operation:
```typescript
const ignoreIfNotExists = edit.options?.ignoreIfNotExists || true
```
Agreed?

Reply:
@UziTech This is what I did: https://github.com/atom-community/atom-languageclient/pull/140/files#diff-9011b4bf132f9c2bff2dfede53c468cea4b1710ce6937a8948a733ca1f109c52R99
Basically the default behavior is: don't throw an error when deleting a file that doesn't exist, except if the server requested `ignoreIfNotExists: false`.